### PR TITLE
Remove pos arg from Perceiver's Pre/Postprocessors

### LINKED ISSUE #15971: PerceiverAudioPreprocessor: forward() missing 1 required positional argument: 'pos'
The [`forward()` method implementation of `PerceiverAudioPreprocessor`](https://github.com/huggingface/transformers/blob/38cc35069c10d153e872162265288263bb7394b7/src/transformers/models/perceiver/modeling_perceiver.py#L3267) seems to be problematic because it has an extra required `pos` argument, whereas all other preprocessors do not (or it is optional). In fact, while it is passed to `_build_network_inputs()`, `pos` is not actually used in that method. Hence, I am not sure why it exists. Moreover, it breaks the default `PerceiverModel` operation, because that class assumes that there only one positional argument called `inputs` [when calling the preprocessor](https://github.com/huggingface/transformers/blob/38cc35069c10d153e872162265288263bb7394b7/src/transformers/models/perceiver/modeling_perceiver.py#L865). This seems to be a serious issue in my opinion because it crashes any instantiation of `PerceiverModel` with a `PerceiverAudioPreprocessor` as `input_preprocessor`.

### PR DESCRIPTION
**Pull request by aielawady**
Fix #15971
@NielsRogge

### CODE REVIEW

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 1]
```diff
@@ -2866,17 +2866,15 @@ def __init__(self, modalities: Mapping[str, PostprocessorType], input_is_dict: b
         self.modalities = nn.ModuleDict(modalities)
         self.input_is_dict = input_is_dict
 
-    def forward(
-        self, inputs: torch.Tensor, pos: Optional[torch.Tensor] = None, modality_sizes=None
-    ) -> Mapping[str, torch.Tensor]:
+    def forward(self, inputs: torch.Tensor, modality_sizes=None) -> Mapping[str, torch.Tensor]:
```
Comment (NielsRogge) [inline src/transformers/models/perceiver/modeling_perceiver.py:2869]: I'm afraid we can't simply remove pos here, due to the fact that this class (`PerceiverMultimodalPostprocessor`) is a public one. Hence, people may have already used this method.

We can only remove pos arguments inside methods which aren't public.

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 2]
```diff
@@ -2897,7 +2895,7 @@ def __init__(self, config: PerceiverConfig, in_channels: int) -> None:
         super().__init__()
         self.classifier = nn.Linear(in_channels, config.num_labels)
 
-    def forward(self, inputs, pos: Optional[torch.Tensor] = None, modality_sizes=None) -> torch.Tensor:
+    def forward(self, inputs, modality_sizes=None) -> torch.Tensor:
```
Comment (NielsRogge) [inline src/transformers/models/perceiver/modeling_perceiver.py:2898]: Same comment here.

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 3]
```diff
@@ -2946,7 +2944,7 @@ def __init__(self, in_channels: int, out_channels: int) -> None:
         super().__init__()
         self.classifier = nn.Linear(in_channels, out_channels)
 
-    def forward(self, inputs: torch.Tensor, pos: Optional[torch.Tensor] = None, modality_sizes=None) -> torch.Tensor:
+    def forward(self, inputs: torch.Tensor, modality_sizes=None) -> torch.Tensor:
```
Comment (NielsRogge) [inline src/transformers/models/perceiver/modeling_perceiver.py:2947]: Same comment here.

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 4]
```diff
@@ -3130,7 +3128,7 @@ def _build_network_inputs(self, inputs: torch.Tensor, pos: torch.Tensor, network
             inputs_with_pos = inputs + pos_enc
         return inputs_with_pos, inputs
 
-    def forward(self, inputs: torch.Tensor, pos: Optional[torch.Tensor] = None, network_input_is_1d: bool = True):
+    def forward(self, inputs: torch.Tensor, network_input_is_1d: bool = True):
```
Comment (NielsRogge) [inline src/transformers/models/perceiver/modeling_perceiver.py:3131]: Same comment here.

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 5]
```diff
@@ -3197,7 +3195,7 @@ def __init__(self, config: PerceiverConfig) -> None:
     def num_channels(self) -> int:
         return self.config.num_labels
 
-    def forward(self, inputs: torch.Tensor, pos: Optional[torch.Tensor] = None, network_input_is_1d: bool = True):
+    def forward(self, inputs: torch.Tensor, network_input_is_1d: bool = True):
```
Comment (NielsRogge) [inline src/transformers/models/perceiver/modeling_perceiver.py:3198]: Same comment here.

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 6]
```diff
@@ -3294,10 +3292,10 @@ def _build_network_inputs(self, inputs, pos):
 
         return inputs_with_pos, inputs
 
-    def forward(self, inputs: torch.Tensor, pos: Optional[torch.Tensor] = None, network_input_is_1d: bool = True):
+    def forward(self, inputs: torch.Tensor, network_input_is_1d: bool = True):
```
Comment (NielsRogge) [inline src/transformers/models/perceiver/modeling_perceiver.py:3295]: Same comment here

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 7]
```diff
@@ -3346,16 +3344,14 @@ def num_channels(self) -> int:
         common_channel_size = max_channel_size + self.min_padding_size
         return common_channel_size
 
-    def forward(
-        self, inputs: Mapping[str, torch.Tensor], pos: Optional[torch.Tensor] = None, network_input_is_1d: bool = True
-    ) -> PreprocessorOutputType:
+    def forward(self, inputs: Mapping[str, torch.Tensor], network_input_is_1d: bool = True) -> PreprocessorOutputType:
```
Comment (NielsRogge) [inline src/transformers/models/perceiver/modeling_perceiver.py:3347]: Same comment here

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 1 above)
Comment (aielawady) [reply src/transformers/models/perceiver/modeling_perceiver.py]: I've reverted the removed `pos` args in the public methods.

File: src/transformers/models/perceiver/modeling_perceiver.py [hunk 8]
```diff
@@ -2924,7 +2922,7 @@ def __init__(self, config: PerceiverConfig, in_channels: int, postproc_type: str
         # Architecture parameters:
         self.classifier = nn.Linear(in_channels, config.samples_per_patch)
 
-    def forward(self, inputs: torch.Tensor, pos: Optional[torch.Tensor] = None, modality_sizes=None) -> torch.Tensor:
+    def forward(self, inputs: torch.Tensor, modality_sizes=None) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (aielawady) [reply src/transformers/models/perceiver/modeling_perceiver.py]: Fixed.

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 4 above)
Comment (aielawady) [reply src/transformers/models/perceiver/modeling_perceiver.py]: Fixed.

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 3 above)
Comment (aielawady) [reply src/transformers/models/perceiver/modeling_perceiver.py]: Fixed.

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 5 above)
Comment (aielawady) [reply src/transformers/models/perceiver/modeling_perceiver.py]: Fixed.

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 6 above)
Comment (aielawady) [reply src/transformers/models/perceiver/modeling_perceiver.py]: Fixed.

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 7 above)
Comment (aielawady) [reply src/transformers/models/perceiver/modeling_perceiver.py]: Fixed.

File: src/transformers/models/perceiver/modeling_perceiver.py (hunk 2 above)
Comment (aielawady) [reply src/transformers/models/perceiver/modeling_perceiver.py]: Fixed.

**Review by NielsRogge — APPROVED** [review]
Thanks for working on this!

**Review by sgugger — APPROVED** [review]
Thanks a lot!

**Pull request merged**